### PR TITLE
Autorefinement: insert points in edge to avoid filter failures

### PR DESCRIPTION
**Pull request by sloriot**
Collect intersection points on edges and in faces to use CDT dedicated functions.
Also add structural filtering with projection traits and restore "classical" projection traits for testing.

### CODE REVIEW

File: Intersections_3/include/CGAL/Intersections_3/internal/Triangle_3_Triangle_3_intersection.h [hunk 1]
```diff
@@ -39,11 +39,15 @@ struct Point_on_triangle
...
   // to always be passed in the same order. For a triangle pqr,
-  // edge 0 is pq, edge 1 qr and edge 2 rp. Point 0 is p, 1 is q and 2 is r.
+  // edge 1 is pq, edge 2 qr and edge 3 rp. Point -1 is p, -2 is q and -3 is r.
+  //
+  //  (0,-) vertex of t2 inside t1
+  //  (-,-) common point of t1 and t2
+  //  (-,0) vertex of t1 inside t2
+  //  (-,+) vertex of t1 on an edge of t2
+  //  (+,+) edge of t1 intersecting edge of t2
+  //  (+,-) vertex of t2 on an edge of t1
```
Comment (albert-github) [inline Intersections_3/include/CGAL/Intersections_3/internal/Triangle_3_Triangle_3_intersection.h:49]: Wouldn't it be  better to have these "magical" numbers as as elements of and `enum` or `enum class` ?

**Comment by sloriot** [comment]
Successfully tested in CGA-6.0-Ic-173 (but TODOs remaining)

**Comment by sloriot** [comment]
There are some runtime errors in CGAL-6.0-Ic-176

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/autorefinement.h [hunk 2]
```diff
@@ -602,9 +680,9 @@ void generate_subtriangles(std::size_t ti,
               {
                 CGAL_AUTOREF_COUNTER_INSTRUCTION(counter.timer6.start();)
                 typename EK::Point_3 pt = typename EK::Construct_planes_intersection_point_3()(
-                  triangles[in_triangle_ids[i]][0], triangles[in_triangle_ids[i]][1],triangles[in_triangle_ids[i]][2],
-                  triangles[in_triangle_ids[j]][0], triangles[in_triangle_ids[j]][1],triangles[in_triangle_ids[j]][2],
-                  triangles[ti][0], triangles[ti][1],triangles[ti][2]);
+                  all_triangle_data[in_triangle_ids[i]].points[0], all_triangle_data[in_triangle_ids[i]].points[1],all_triangle_data[in_triangle_ids[i]].points[2],
```
Comment (afabri) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/autorefinement.h:683]: With VC2022 and the call `test_autorefinement.exe data-autoref/four_cubes.off 12 94 184 1 78 0 106 352`   I get a segfault in the parallel execution.   My guess (only based on gut feeling) is  that while a thread is in this functor another thread executes the exact construction of one of the input points

Comment (sloriot) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/autorefinement.h:683]: I can produce an error locally on the same case (need several runs). Thanks for the pointer. Investigating...

**Comment by sloriot** [comment]
Successfully tested in CGAL-6.0-Ic-181

**Pull request merged**